samtools: add page

ewels opened the pull request:
First stab at adding a new page for [samtools](http://www.htslib.org/doc/samtools.html). See #1084.

ewels commented:
It's kind of a fun game trying to condense stuff down into such a short list of examples 😄  Note that there are a bunch of subcommands that I didn't add, plus some have so many options that they could probably have their own _tldr_ page (eg. `samtools view`).

I've done all of this based on the commands that I use most myself, others may have different opinions.

waldyrious commented on pages/common/samtools.md:
```diff
@@ -0,0 +1,28 @@
+# samtools
+
+> Tools for handling high-throughput sequencing (genomics) data.
+> Reading/writing/editing/indexing/viewing of data in SAM/BAM/CRAM format.
```
Minor nitpick, but I'd prefer this sentence to have a subject.

ewels replied:
Yup, this is me being lazy and copying + pasting again.

waldyrious commented on pages/common/samtools.md:
```diff
@@ -0,0 +1,28 @@
+
+- Convert a SAM file to BAM (use - for stdin):
+
+`samtools view -bS {{input.sam}} > {{output.bam}}`
```
is the S an argument to the -b option, or a separate option? If so, what do those letters stand for?

ewels replied:
`-S` is a separate option. It specifies `SAM` input format. In the latest versions of samtools this flag is actually ignored, the format is autodetected from the input filename extension. However, a range of samtools versions are still common so I typically include it in scripts for backwards compatibility.

`-b` specifies that the output should be in `BAM` format instead of the default `SAM` for `stdout`. This isn't needed if using `-o {{output.bam}}` (autodetected by file extension again).

waldyrious commented on pages/common/samtools.md:
```diff
@@ -0,0 +1,28 @@
+
+- Print a region of a file, include the file header:
+
+`samtools view -h {{input}} chr:start-end > {{output.sam}}`
```
What does "chr" stand for?

ewels replied:
Chromosome. I would hope that this is pretty common knowledge for anyone using the tool, but I could extend it to `chromosome:start-end`.

waldyrious commented:
The condensing challenge isn't easy, indeed. One of the things that helps is using the example descriptions to explain what the options mean, and thus create a mental model that would help infer other combinations of options. We're actually considering expanding the maximum number of examples to 8, but in this case, have you considered the option of splitting these commands into one page per subcommand, as is done with git?

ewels commented:
I did think about expanding these into multiple pages, but I wanted to remain consistent. Not all commands have lots of options and probably don't deserve their own files. I think with a bit of tweaking based on your suggestions that this is probably a happy middle ground for most people - covering the most commonly used commands without going into too much depth.

ewels commented:
Thanks again for the suggestions - update pushed with better descriptions (explaining flags more explicitly). I also changed the example commands a little to demonstrate typical unix piping, as this is often how you use the tool. Finally I added a couple more command examples to push up 8.

waldyrious commented on pages/common/samtools.md:
```diff
@@ -0,0 +1,36 @@
+# samtools
+
+> Tools for handling high-throughput sequencing (genomics) data.
```
would "sequencing (genomics) data" work better as "gene sequencing data" (or some variation thereof if that isn't technically correct)?

ewels replied:
_"high-throughput sequencing"_ is a commonly used term for this (often abbreviated to HTS), so those words go together quite nicely. I added _(genomics)_ to help clarify. If I'm honest I can't think of any alternatives that sound better to my ear..

waldyrious commented on pages/common/samtools.md:
```diff
@@ -0,0 +1,36 @@
+
+`samtools view -bS {{input.sam}} > {{output.bam}}`
+
+- Take input from stdin (-) and output a region, including the file header (-h). Prints to stdout:
```
To clarify, here it's ok to mention the link `stdin` --> `-`, because it's not obvious.

waldyrious commented on pages/common/samtools.md:
```diff
@@ -0,0 +1,36 @@
+
+`samtools flagstat {{sorted_input}}`
+
+- Count alignments to each index (chromosome / contig):
```
Could "contig" be expanded into its full form?

ewels replied:
"Contig" is the full form :) https://en.wikipedia.org/wiki/Contig

waldyrious commented on pages/common/samtools.md:
```diff
@@ -0,0 +1,36 @@
+
+- Merge multiple files:
+
+`samtools merge {{output}} {{input_1}} [{{input_2}}..{{input_n}}]`
```
I'm wondering if `[{{input_2}}...]` would be clearer. @agnivade, any thoughts?

ewels commented on pages/common/samtools.md:
```diff
@@ -0,0 +1,36 @@
+
+- Print alignment statistics about a file:
+
+`samtools flagstat {{sorted_input}}`
```
It's a numerical value assigned to a DNA read when it's aligned to a reference genome. There are several of them and they can be combined. There's a nice tool here to decode them, which should give an idea of what they look like: https://broadinstitute.github.io/picard/explain-flags.html

ewels commented:
Thanks again! I agree with pretty much all of your suggestions, have just pushed a quick update to address your comments. Let me know what you think..

waldyrious commented:
LGTM! Great contribution, thanks!

agnivade commented on pages/common/samtools.md:
```diff
@@ -0,0 +1,36 @@
+
+`samtools sort {{input}} -o {{output.bam}}`
+
+- Index a sorted BAM file (creates {{sorted_input.bai}}):
```
shouldn't this be `sorted_input.bam` ?

ewels replied:
The input file can be `.bam` or `.cram` format, but yes it probably should be `sorted_input.bam` as I specify _BAM_ in the description (plus this is by far the most common). 👍

ewels replied:
..(updated)

ewels replied:
Oh I see, nope! That's meant to be `bai` :) Short for for _bam index_. Though actually you are correct that there's a typo - it should be `sorted_input.bam.bai` now that we have the input file as `sorted_input.bam`

ewels replied:
ps. If you're curious, here is the full description of the command output:

> If an output filename is given, the index file will be written to _out.index_. Otherwise, for a CRAM file _aln.cram_, index file _aln.cram.**crai**_ will be created; for a BAM file _aln.bam_, either _aln.bam.**bai**_ or _aln.bam.**csi**_ will be created, depending on the index format selected.

..confused yet? ;) 90% of people will be using this command with `.bam` files though, so I think it's clearest to make the example explicitly for that, creating `.bam.bai` files as we now have.

agnivade commented:
LGTM
